flaml/automl/model.py

```diff
@@ -386,6 +418,342 @@ def config2params(self, config: dict) -> dict:
+        return {
+            "numIterations": {
+                "domain": tune.lograndint(lower=4, upper=upper),
+                "init_value": 4,
+                "low_cost_init_value": 4,
+            },
+            "numLeaves": {
+                "domain": tune.lograndint(lower=4, upper=upper),
+                "init_value": 4,
+                "low_cost_init_value": 4,
+            },
+            "minDataInLeaf": {
+                "domain": tune.lograndint(lower=2, upper=2**7 + 1),
+                "init_value": 20,
+            },
+            "learningRate": {
+                "domain": tune.loguniform(lower=1 / 1024, upper=1.0),
+                "init_value": 0.1,
+            },
+            "log_max_bin": {  # log transformed with base 2
+                "domain": tune.lograndint(lower=3, upper=11),
+                "init_value": 8,
+            },
+            "featureFraction": {
+                "domain": tune.uniform(lower=0.01, upper=1.0),
+                "init_value": 1.0,
+            },
+            "lambdaL1": {
+                "domain": tune.loguniform(lower=1 / 1024, upper=1024),
+                "init_value": 1 / 1024,
+            },
+            "lambdaL2": {
+                "domain": tune.loguniform(lower=1 / 1024, upper=1024),
+                "init_value": 1.0,
+            },
+        }
+
+    def config2params(self, config: dict) -> dict:
+        params = super().config2params(config)
+        if "log_max_bin" in params:
+            params["maxBin"] = (1 << params.pop("log_max_bin")) - 1
+        return params
+
+    @classmethod
+    def size(cls, config):
+        num_leaves = int(
+            round(config.get("numLeaves") or 1 << config.get("maxDepth", 16))
+        )
+        n_estimators = int(round(config["numIterations"]))
+        return (num_leaves * 3 + (num_leaves - 1) * 4 + 1.0) * n_estimators * 8
+
+    def __init__(self, task="binary", **config):
+        super().__init__(task, **config)
+        if "verbose" in self.params:
+            self.params.pop("verbose")
+        if "n_jobs" in self.params:
+            self.params.pop("n_jobs")
+        err_msg = (
+            "SynapseML is not installed. Please refer to [SynapseML]"
+            + "(https://github.com/microsoft/SynapseML) for installation instructions."
+        )
+        if "regression" == task:
+            try:
+                from synapse.ml.lightgbm import LightGBMRegressor
+            except ImportError:
+                raise ImportError(err_msg)
+
+            self.estimator_class = LightGBMRegressor
+            self.estimator_params = ParamList_LightGBM_Regressor
+        elif "rank" == task:
+            try:
+                from synapse.ml.lightgbm import LightGBMRanker
+            except ImportError:
+                raise ImportError(err_msg)
+
+            self.estimator_class = LightGBMRanker
+            self.estimator_params = ParamList_LightGBM_Ranker

```

Comment:
What are these used for? `estimator_params`.

Reply:
It's the list of all the parameters supported by the model. As flaml supports mixing parameters of automl settings and parameters of models. We need to pop out non-model-params before passing kwargs to model fit function, otherwise the model fit will raise error.

For instance, in below example, label is for automl, labelCol and isUnbalance is for actual model fit. Other kwargs in **settings may also contain invalid params for the model.
```
automl.fit(
    dataframe=df,
    label="Bankrupt?",
    labelCol="Bankrupt?",
    isUnbalance=True,
    **settings,
)
```

Reply:
Where is `self.estimator_params` used?

Reply:
https://github.com/thinkall/FLAML/blob/63727bc817e0d998db8d505148517c8563057dc2/flaml/automl/model.py#L682-L684
```
        for k in list(_kwargs.keys()):
            if k not in self.estimator_params:
                _kwargs.pop(k)
```

Reply:
Not all the arguments in settings are passed to the estimator's fit() function. The above code is unnecessary. Did I miss something?

Reply:
Users may pass in invalid params, in the above example, `isUnbalance` is only valid for Classifier, if we don't pop it out in other models, the code will raise an error. And some configs like `n_jobs`, `verbose` will also be passed in. In the future, I think more configs could be passed in. Since these synapseML models will just raise errors with invalid params, I added this piece of code to make it more robust.

To me, it's more like a trade-off between the code and user experiences.

Reply:
I often see this warning message, `[LightGBM] [Warning] Find whitespaces in feature_names, replace with underlines`, which is also an example that encourages me to sacrifice a little bit performance to make the code more robust.

Reply:
When users pass in invalid params, how about issuing a warning?
For `n_jobs`, it is passed to `self.params` and popped in line 629-630 (actually, I suggest move line 629-630 into the `config2params()` function). It's not in `_kwargs`.
For `verbose`, it shouldn't be part of `self.params`, so line 627-628 is unnecessary I think.
Is the warning message related to the code snippet we are discussing?

Reply:
Removed `verbose`, moved `n_jobs` to `config2params()`.

The `[LightGBM] [Warning] Find whitespaces in feature_names, replace with underlines` happens when using `lgbm_spark`, but not happens when using `lgbm`. And I'm not handling this in flaml.

For invalid params passed into synapseML-LightGBM (the model behind `lgbm_spark`), it will raise an error, so I've to handle the params in flaml before passing to synapseML-LightGBM. The other way is to raise the error and let the user correct their inputs.

Reply:
I'm OK with both ways: issue a warning and remove; or raise an error. Just don't remove silently.

Reply:
> I'm OK with both ways: issue a warning and remove; or raise an error. Just don't remove silently.

Tried locally, found that `estimator_list`, `verbose` were both sent to fit function. As there could be these valid params for AutoML, but not for model fitting be passed into fit function. There could be too many unnecessary warning issues. And we couldn't let the model raise errors as `verbose`/`estimator_list` are both valid params for automl.  
```
[flaml.automl.logger: 03-17 16:41:32] {671} WARNING - [SynapseML Warning] Invalid parameter estimator_list is ignored.
[flaml.automl.logger: 03-17 16:41:32] {671} WARNING - [SynapseML Warning] Invalid parameter verbose is ignored.
```

Reply:
`estimator_list` and `verbose` are not supposed to be passed to the fit function because they are specified kwargs in `AutoML.fit()` and shouldn't be part of `fit_kwargs`. There is a bug I suppose.

Reply:
> `estimator_list` and `verbose` are not supposed to be passed to the fit function because they are specified kwargs in `AutoML.fit()` and shouldn't be part of `fit_kwargs`. There is a bug I suppose.

It happens when using `retrain_from_log`. I'm not sure if this is a bug or not. But we can remove these settings before we do `retrain_from_log`. And remove invalid parameters is still needed for SparkLGBMEstimator.

An example is `test_spark_synapseml_rank()` in `test_0sparkml.py`.

Reply:
When using `retrain_from_log`, `estimator_list` and `verbose` are supposed to be removed from the arguments. That's the current design. If we don't plan to change that design in this PR, then `test_spark_synapseml_rank()` should be modified.

Reply:
Yes, I have removed verbose and estimator_list from the arguments in test_spark_synapseml_rank().